### ISSUE TITLE
Helm chart improvements

### ISSUE THREAD

**Issue by grzesuav** [metacontroller/metacontroller#302]
Recently, thanks to @sathieu  we have helm chart, which can be improved in following way :

1. Add customization to arguments passed to metacontroller
2. When releasing, make chart bundle and attach to release resources as `assets` on https://github.com/metacontroller/metacontroller/releases page
3. Single source of truth for manifests, currently helm and manifests are separated, ideally manifests with different versions - `dev`/`debug`/`production` should be a result of helm chart generation with different input values happening on release time
4. CRD generation should point to helm chart once 3. is done

**Comment by sathieu** [comment]
> 2. When releasing, make chart bundle and attach to release resources as assets on https://github.com/metacontroller/metacontroller/releases page

Having a chart index would be better. As an example, grafana [does this](https://github.com/grafana/helm-charts/blob/main/.github/workflows/release.yaml) using github pages.

**Comment by grzesuav** [comment]
> > 1. When releasing, make chart bundle and attach to release resources as assets on https://github.com/metacontroller/metacontroller/releases page
> 
> Having a chart index would be better. As an example, grafana [does this](https://github.com/grafana/helm-charts/blob/main/.github/workflows/release.yaml) using github pages.

@sathieu can you post here a link to what chart index it ? I am noob in case of helm

**Comment by grzesuav** [comment]
Actually I think they publish helm chart  release page - https://github.com/grafana/helm-charts/releases

**Comment by sathieu** [comment]
@grzesuav The charts index is at https://grafana.github.io/helm-charts/index.yaml. It can be generated with `helm repo index`, see [here](https://helm.sh/docs/topics/chart_repository/) for generating it with shell script. I chose grafana charts as examples because they use github workflows.

**Comment by grzesuav** [comment]
thanks, will look :+1:

**Comment by sathieu** [comment]
See also https://github.com/marketplace/actions/helm-chart-releaser.

**Comment by VictorBac** [comment]
I just found this thread, I made this earlier today because I needed it:
https://github.com/metacontroller/metacontroller/pull/364

This should correspond to your point number 1

**Comment by orf** [comment]
This would be very useful, it's quite inconvenient to install this right now.

**Comment by sathieu** [comment]
@orf Could you propose a PR here to implement this here?

**Comment by orf** [comment]
Sure @sathieu, I've made #382.

**Comment by grzesuav** [comment]
I closed #382 , it was using github pages to keep the charts. it is busy with serving documentation

**Comment by grzesuav** [comment]
Would be nice to use github packages for that - https://github.com/github/roadmap/issues/120 but still unimplemented

**Comment by grzesuav** [comment]
hmm actually seems possible, need to check that - https://www.visualstudiogeeks.com/github/publish-helm-3-charts-to-gcr

**Comment by sathieu** [comment]
@grzesuav Move to gitlab 😉 !

https://docs.gitlab.com/ee/user/packages/helm_repository/
https://github.com/failfast-ci/failfast-api

**Comment by grzesuav** [comment]
@sathieu @orf @VictorBac @mjsmith1028 do you see any outstanding improvements to helm charts ? Currently they are shipped as github packages and described in docs.

**Comment by mjsmith1028** [comment]
Hey @grzesuav 

I have two things that I can think of at the moment.

1. See https://github.com/metacontroller/metacontroller/pull/484#discussion_r837469764 . I think we should simply use `commandArgs` to provide all cli args instead of the current zap and discovery overrides. This would be a breaking change but would simplify the chart and allow the the dlv debug image to be ran through the helm chart. In addition, allow for the command to be overridden to allow the `dlv` command instead of `/usr/bin/metacontroller`. If this makes sense to you, I can add those changes in that PR.

2. I would like to have some toggle in the values file to allow for the ServiceMonitor and additional setup to expose prometheus metrics. Similar to the example https://github.com/metacontroller/metacontroller/tree/master/examples/prometheus-monitoring

**Comment by sathieu** [comment]
@grzesuav I think having the Helm repository as HTTPs (i.e. not OCI) is more standard. Also the  current name of the chart (` metacontroller-helm`) is not standard.

However, this is not a big deal, and I'm able to [workaround this easily](https://gitlab.com/kubitus-project/kubitus-ingress-operator/-/merge_requests/72/diffs?commit_id=f6889bbf5ff09810329c991eba0ba501f04e8e2b).

**Comment by sathieu** [comment]
@grzesuav Also related, some improvements can be done on the image to run as non-root.

**Comment by grzesuav** [comment]
> @grzesuav I think having the Helm repository as HTTPs (i.e. not OCI) is more standard.

True, however OCI was added as supported repository format in helm, Ig you could point to alternative page to deploying charts I can try to add it as alternative

> Also the current name of the chart (` metacontroller-helm`) is not standard.

True, it is because `metacontroller` is used to store docker image on hcr.io

> 
> However, this is not a big deal, and I'm able to [workaround this easily](https://gitlab.com/kubitus-project/kubitus-ingress-operator/-/merge_requests/72/diffs?commit_id=f6889bbf5ff09810329c991eba0ba501f04e8e2b).

I am not able to tell what this workaround is actually doing, what was the original issue ?

**Comment by sathieu** [comment]
> True, however OCI was added as supported repository format in helm, Ig you could point to alternative page to deploying charts I can try to add it as alternative

There is a github action: https://github.com/helm/chart-releaser-action

> I am not able to tell what this workaround is actually doing, what was the original issue ?

The workaround is about avoiding `metacontroller-helm` in values when metacontroller is a subchart (this was a breaking change, and also not very explicit).

**Comment by grzesuav** [comment]
> > True, however OCI was added as supported repository format in helm, Ig you could point to alternative page to deploying charts I can try to add it as alternative
> 
> There is a github action: https://github.com/helm/chart-releaser-action
> 
https://github.com/metacontroller/metacontroller/issues/302#issuecomment-1009409034

**Comment by mjsmith1028** [comment]
> @grzesuav Also related, some improvements can be done on the image to run as non-root.

I have added this on PR https://github.com/metacontroller/metacontroller/pull/534. Thanks!
